interp/src/primitives/combinational.rs

```diff
@@ -67,37 +74,187 @@ comb_primitive!(StdNot[WIDTH](r#in: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdFpSsub[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdFpSmultPipe[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
+});
+comb_primitive!(StdFpSdivPipe[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
+    todo!()
 });
 
 // ===================== Shift Operations ======================
 comb_primitive!(StdLsh[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
-    let mut tr = left.vec.clone();
-    tr.shift_right(right.as_u64() as usize);
-    Value { vec: tr }.into()
+    //to avoid the casting overflow,
+        //we know that [left], [right], and [self]
+        //are capped at bitwidths as large as largest u64 (2^64 - 1 = 1.84 * 10^19 ...)
+        //so, if [self] has a width greater than 64,
+        //and the 65th index is a 1, we just automatically return a 0 of the
+        //appropriate bitwidth!
+
+        if WIDTH > 64 {
+            //check if right is greater than or equal to  2 ^ 64
+            let r_vec = &right.vec;
+            for (index, bit) in r_vec.iter().by_ref().enumerate() {
+                if (index >= 64) & *bit {
+                    return Value::zeroes(WIDTH as usize).into();
+                }
+            }
+        }
+
+        //but that's not the only problem. we can't let [place] get to
+        //2^64 or above. However the right value couldn't have even been specified
+        //to be greater than or equal to 2^64, because it's constrained by u64.
+        //so instead of incrementing [place], just calculate place, but only
+        //when [bit] is 1, which can only be true for bits below the 65th (first
+        // bit is 2^0)
+
+        let mut tr = BitVec::new();
+        //first push the requisite # of zeroes
+        for (index, bit) in right.vec.iter().by_ref().enumerate() {
+            if *bit {
+                //not possible for bit to be 1 after the 64th bit
+                for _ in 0..u64::pow(2, index as u32) {
+                    if tr.len() < WIDTH as usize {
+                        tr.push(false);
+                    }
+                    //no point in appending once we've filled it all with zeroes
+                }
+            }
+        }
+        //then copy over the bits from [left] onto the back (higher-place bits) of
+        //[tr]. Then truncate, aka slicing off the bits that exceed the width of this
+        //component
+        let mut to_append = left.clone().vec;
+        tr.append(&mut to_append);
+        tr.truncate(WIDTH as usize);
+        let tr = Value { vec: tr };
+        assert_eq!(tr.width(), WIDTH);
+        //sanity check the widths
+        tr.into()
 });
 comb_primitive!(StdRsh[WIDTH](left: WIDTH, right: WIDTH) -> (out: WIDTH) {
-    let mut tr = left.vec.clone();
-    tr.shift_left(right.as_u64() as usize);
-    Value { vec: tr }.into()
+    //remove [right] bits from index 0
+        //extend to proper size
+
+        //same check as in LSH
```

Comment:
Makes me wonder if it's possible to factor out into a function with a flag that says whether to shift left or right… maybe!